sysl2/sysl/ints.go

```diff
@@ -0,0 +1,253 @@
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+type IntegrationParam struct {
+	m            *sysl.Module
+	excludes     StrSet
+	passthroughs StrSet
+	commonEndpts StrSet
+	highlights   StrSet
+	apps         StrSet
+	deps         map[string]*AppDependency
+}
+
+type ApplicationCollectionElement struct {
+	apps StrSet
+}
+
+type IntegrationStatementElement struct {
+	appname string
+	epname  string
+	stmts   []*sysl.Statement
+}
+
+func MakeIntegrationParam(mod *sysl.Module, excludes, passthroughs, highlights StrSet) *IntegrationParam {
+	return &IntegrationParam{
+		m:            mod,
+		excludes:     excludes,
+		passthroughs: passthroughs,
+		commonEndpts: MakeStrSet(".. * <- *", "*"),
+		highlights:   highlights,
+		apps:         StrSet{},
+		deps:         map[string]*AppDependency{},
+	}
+}
+
+func MakeApplicationCollectionElement(apps StrSet) *ApplicationCollectionElement {
+	return &ApplicationCollectionElement{apps}
+}
+
+func MakeIntegrationStatementElement(appname, epname string, stmts []*sysl.Statement) *IntegrationStatementElement {
+	return &IntegrationStatementElement{appname, epname, stmts}
+}
+
+func (v *IntegrationParam) collectApplicationDependencies(e *ApplicationCollectionElement) {
+	for app := range e.apps {
+		for epname, endpt := range v.m.GetApps()[app].GetEndpoints() {
+			i := MakeIntegrationStatementElement(app, epname, endpt.GetStmt())
+			v.handleStatement(i)
+		}
+	}
+}
+
+func (v *IntegrationParam) handleStatement(i *IntegrationStatementElement) {
+	for _, stmt := range i.stmts {
+		switch t := stmt.GetStmt().(type) {
+		case *sysl.Statement_Call:
+			targetName := getAppName(t.Call.GetTarget())
+			if ok := v.addAppDependency(i.appname, i.epname, targetName, t.Call.Endpoint, stmt); !ok {
+				continue
+			}
+			v.addEndpoint(targetName, t.Call.Endpoint)
+		case *sysl.Statement_Action, *sysl.Statement_Ret:
+			continue
+		case *sysl.Statement_Cond:
+			e := MakeIntegrationStatementElement(i.appname, i.epname, t.Cond.GetStmt())
+			v.handleStatement(e)
+		case *sysl.Statement_Loop:
+			e := MakeIntegrationStatementElement(i.appname, i.epname, t.Loop.GetStmt())
+			v.handleStatement(e)
+		case *sysl.Statement_LoopN:
+			e := MakeIntegrationStatementElement(i.appname, i.epname, t.LoopN.GetStmt())
+			v.handleStatement(e)
+		case *sysl.Statement_Foreach:
+			e := MakeIntegrationStatementElement(i.appname, i.epname, t.Foreach.GetStmt())
+			v.handleStatement(e)
+		case *sysl.Statement_Group:
+			e := MakeIntegrationStatementElement(i.appname, i.epname, t.Group.GetStmt())
+			v.handleStatement(e)
+		case *sysl.Statement_Alt:
+			for _, choice := range t.Alt.GetChoice() {
+				e := MakeIntegrationStatementElement(i.appname, i.epname, choice.GetStmt())
+				v.handleStatement(e)
+			}
+		default:
+			panic("No statement!")
+		}
+	}
+}
+
+func (v *IntegrationParam) addAppDependency(sourceApp, sourceEndpt, targetApp, targetEndpt string, stmt *sysl.Statement) bool {
+	if len(v.apps) == 0 {
+		v.apps = v.apps.Union(v.highlights)
```

Comment:
I think it would get a problem. If do like this, when the value of v.apps were changed, the v.highlights would also be changed. However, v.highlights should be kept with fixed values.

Reply:
Actually v.apps will only be empty the first time and this statement will only be executed exactly once.. 
To me this code is in the wrong place. This looks like initialisation, but v.apps is initialized as empty str set.